### DISCUSSION TITLE
Problem casting ringbuffer_sink_st

### DISCUSSION THREAD

**Discussion by dradic2** [gabime/spdlog#2411] [Q&A] [answered]
Hello,

I am using the ring buffer sink in a multiple sink configuration. Setting up the logger in main and using it somewhere else.

    secondSink = std::make_shared<spdlog::sinks::ringbuffer_sink_st>(50);
    secondSink->set_level(spdlog::level::level_enum::trace);

    sinks.push_back(firstSink);
    sinks.push_back(secondSink);

    auto logger = std::make_shared<spdlog::logger>("my_logger", begin(sinks), end(sinks));
    logger->set_level(spdlog::level::level_enum::trace);`
    spdlog::register_logger(logger);

This seems to be working as I can see the log messages in the buffer.

The problem is, how cast the sink to ring_buffer_st to be able to use last_formatted(10)

    auto logger = spdlog::get("my_logger");
    auto sink = (logger->sinks().back());
    auto ringBufferSink = (std::make_shared<spdlog::sinks::ringbuffer_sink_st>)(logger->sinks().end());
    ringBufferSink->last_formatted(10);

I get the following error

    error: no matching function for call to ‘spdlog::sinks::ringbuffer_sink<spdlog::details::null_mutex>::ringbuffer_sink(__gnu_cxx::__normal_iterator<std::shared_ptr<spdlog::sinks::sink>*, std::vector<std::shared_ptr<spdlog::sinks::sink> > >)’

If I don't try casting, I can't use last_formatted()

Help would be appreciated!

Thanks,
Dusan

**Comment by tt4g** [comment, marked as answer]
Don't use C style cast for casting `std::shared_ptr<T>`. Must use `std::dynamic_pointer_cast` instead (this is a problem with the use of `std::shared_ptr<T>`).
See: https://en.cppreference.com/w/cpp/memory/shared_ptr/pointer_cast

```diff
- auto ringBufferSink = (std::make_shared<spdlog::sinks::ringbuffer_sink_st>)(logger->sinks().end());
+ auto ringBufferSink = std::dynamic_pointer_cast<spdlog::sinks::ringbuffer_sink_st>(logger->sinks().back());
+ if (ringBufferSink) {
+      ringBufferSink->last_formatted(10);
+ } else {
+    // Failed to cast.
+ }

**Reply by dradic2** [reply]
Tank you!
